lib/pacer.go

```diff
@@ -0,0 +1,170 @@
+package vegeta
+
+import (
+	"fmt"
+	"math"
+	"time"
+)
+
+// A Pacer defines the rate of hits during an Attack by
+// returning the duration an Attacker should wait until hitting
+// the next Target. If the second return value is true, the
+// attack will terminate.
+type Pacer interface {
+	Pace(elapsedTime time.Duration, elapsedHits uint64) (sleep time.Duration, stop bool)
+}
+
+// A PacerFunc is a function adapter type that implements the
+// Pacer interface.
+type PacerFunc func(time.Duration, uint64) (time.Duration, bool)
+
+// Pace implements the Pacer interface.
+func (f PacerFunc) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	return f(elapsedTime, elapsedHits)
+}
+
+// Rate sends a constant rate of hits to the target.
+type Rate struct {
+	Freq int           // Frequency (number of occurrences) per ...
+	Per  time.Duration // Time unit, usually 1s
+}
+
+// Pace determines the length of time to sleep until the next hit is sent.
+func (r Rate) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	if r.Per <= 0 || r.Freq <= 0 {
+		// If Rate configuration is invalid, stop the attack.
+		return 0, true
+	}
+	interval := uint64(r.Per.Nanoseconds() / int64(r.Freq))
+	delta := time.Duration((elapsedHits + 1) * interval)
+	// Zero or negative durations cause time.Sleep to return immediately.
+	return delta - elapsedTime, false
+}
+
+// String returns a pretty-printed description of the rate, e.g.:
+//   Rate{1 hits/1s} for Rate{Freq:1, Per: time.Second}
+func (r Rate) String() string {
+	return fmt.Sprintf("Rate{%d hits/%s}", r.Freq, r.Per)
+}
+
+func (r Rate) HitsPerNs() float64 {
+	return float64(r.Freq) / float64(r.Per)
+}
+
+var _ Pacer = Rate{}
+
+const (
```

Comment:
What are these constants used for? It'd be good to explain in docs.

Reply:
Each one individually.

Reply:
Will do. They're useful constant values for the `StartAt` `SinePacer` struct field.